### PR TITLE
makefiles/docker.inc.mk: Pin riotbuild version with BUILD_IN_DOCKER=1

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

We had a number of issues being reported that in the end were caused by building with a version of riotbuild incompatible with the source.

This commits adds a check that the used image is in the listed of known to work versions, unless `DOCKER_IMAGE` is not the default. Users overwriting `DOCKER_IMAGE` are expected to know what they are doing.

This also changes behavior in case no docker image is present: It will error and tell the use to use docker pull to get the image (and how to get it). **This is a change in behavior.**

I believe the new behavior of not downloading the image when not present is more consistent. E.g. users are otherwise mislead to the assumption that the image is kept in sync automatically, which never was the case. 

### Testing procedure

Compile an app with `BUILD_IN_DOCKER=1`. It should be have as follows:

1. The image is not locally available
    - It should fail and tell the user to pull it (and how to pull it)
2. The image is locally available, but the version is not in the list of known-good images
    - It should fail and tell the user the image is not compatible. It should hint that pulling the latest version may fix it
3. The image is locally available in the version that is compatible with the source
    - It should build as before
4. An alternative image has been specified with `DOCKER_IMAGE` e.g. by hash or name
    - The check for compatibility is skipped

### Issues/PRs references

Alternative to https://github.com/RIOT-OS/RIOT/pull/20470

Depends on and includes:

- [ ] https://github.com/RIOT-OS/RIOT/pull/20721

### CODE REVIEW

**Review by mguetschow — DISMISSED** [review]
LGTM in general, although a proper semver versioning as suggested by @kaspar030 would probably be the cleaner solution. Having a hardcoded hash in here requires us to remember updating it whenever breaking changes happen to RIOT sources.

In any case, I would add a sentence to the documentation explicitly stating this behavior for the default docker image.

File: makefiles/docker.inc.mk
Comment (kfessel) [inline makefiles/docker.inc.mk:15]:
```suggestion
# special version "none" refers to "not existing".
```

**Comment by mguetschow** [comment]
The Docker container got updated with the [last merge](https://github.com/RIOT-OS/riotdocker/commits/master/), so we would need to update the whitelist already: https://hub.docker.com/r/riot/riotbuild/tags. I'm wondering if it wouldn't be better to enforce a single specific Docker image at any point in time to avoid having an ever-growing whitelist that would somehow need to be checked for backwards compatibility?

@maribu would you mind adding a sentence to the documentation so we can get this in before the next release?

**Comment by mguetschow** [comment]
A suggestion on how to move this forward:

1. only have a single, tested and guaranteed-to-work riotbuild docker image version for every RIOT commit (i.e., change `COMPATIBLE_RIOTBUILD_VERSIONS` to a single `TESTED_RIOTBUILD_VERSION`)
1. do not abort the build when the downloaded docker image does not match (no breakage in trains), but instead issue a prominent warning, suggesting `docker pull docker.io/riot/riotbuild:sha256@b6af289bf59df73766d685d0df8b8aa98d4fd07b19457699a8a946d204a3405d` (see [docker docs](https://docs.docker.com/reference/cli/docker/image/pull/#pull-an-image-by-digest-immutable-identifier)) and mentioning that things may break with this docker image version

How does that sound to you @maribu ? If you don't find the time for this right now, I can also take over the PR if you want me to.

**Comment by maribu** [comment]
I wonder if it is just better to set the docker image to `docker.io/riot/riotbuild:<hash>` instead of `docker.io/riot/riotbuild:latest` and document in riotbuild to update that hash in RIOT on every bump.

Recalling the last few bumps of riotbuild there was only a single one (adding python-serial) would not have needed to update the docker image to remain bug-for-bug and overflow-for-overflow compatible with the CI.

**Comment by mguetschow** [comment]
> I wonder if it is just better to set the docker image to `docker.io/riot/riotbuild:<hash>` instead of `docker.io/riot/riotbuild:latest` and document in riotbuild to update that hash in RIOT on every bump.

Sounds good to me, but I still think the proposed behavior changes of this PR (no automatic download on first use, warning on wrong riotbuild version) make sense.

(I think it needs to be `docker.io/riot/riotbuild:sha256@<hash>` as written above)

**Comment by maribu** [comment]
> but I still think the proposed behavior changes of this PR (no automatic download on first use, warning on wrong riotbuild version) make sense.

I think the new behavior is even better: When the required docker image is locally available, it will never pull. But if not, it will pull automatically. That would allow bisecting with `BUILD_IN_DOCKER=1` to work like a charm even if toolchain and source need a matching version.

Note: In a train in Brandenburg without the docker image available, one could still run `make DOCKER_IMAGE=docker.io/riot/riotbuild:latest` to just go with the latest locally available instead of a compatible one and hope for the best.

**Comment by mguetschow** [comment]
Thanks for picking this up again!

I've just tried the current changeset locally, and it seems `docker run (...) $image:sha256@$hash` doesn't work with my setup (`Docker version 20.10.24+dfsg1, build 297e128`). Instead, it should be `docker run (...) $hash` or `docker run sha256:$hash`, see here: https://github.com/docker/cli/issues/2815

This makes the build fail with the not very user-friendly error message:

    docker: Error response from daemon: No such image: sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b2.

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:3]:
```suggestion
# a random RIOT merge commit, `make BUILD_IN_DOCKER=1` should always succeed.
```

**Comment by mguetschow** [comment]
> > No such image
> 
> That makes sense. The `DROP ME!!! TESTING CI ONLY` commit changes the hash to something invalid, so that I can see that the static test would indeed complain when someone would update the CI but forget to bump the hash in `makefiles/docker.inc.mk`.

Yeah sorry, this also happens when using the right hash with 3 at the end. The point is just that docker tries to find the image locally, but when failing to do so, it will just report that it doesn't exist instead of pulling it (since it's missing the docker image location information).

**Comment by kfessel** [comment]
maybe you can make the Brandenburg case a little more accessible ( and have it produce a warning message)

**Comment by maribu** [comment]
OK, at least the CI integration seems to work as expected now:

image

**Comment by maribu** [comment]
I also fail to pull an image from dockerhub by SHA256 with both podman and docker. For now, this just goes for `docker pull ...:latest` instead, which would solve the typical issue of the local image out of date.

**Comment by mguetschow** [comment]
Apparently there are two different hashes involved: The local image ID which can be used to refer to the image, and the repo digest which is the one that is [shown on Dockerhub](https://hub.docker.com/layers/riot/riotbuild/latest/images/sha256-75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1?context=explore).

One can see both with `docker inspect <image>` in as in `docker inspect --format='{{index .RepoDigests 0}}' riot/riotbuild:latest` and `docker inspect --format='{{index .Id}}' riot/riotbuild:latest`.

File: makefiles/docker.inc.mk
Comment (maribu) [reply makefiles/docker.inc.mk:37]: Good catch. (Locally I have `DOCKER=podman`, so I wouldn't notice.)

**Comment by maribu** [comment]
Indeed! I now have just added both digest to the Makefile and let the CI verify both are up to date.

It is IMO pretty odd to have one ID to identify an image to run, and an unrelated ID to identify the same image for downloading. But anyway, seems to work now.

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:45]: Why do we need this line?

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:46]:
```suggestion
        endif
    endif
```

File: makefiles/docker.inc.mk
Comment (maribu) [reply makefiles/docker.inc.mk:48]: I added a comment to that line.

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:44]:
```suggestion
        $(info Building with latest available riotbuild image. You can pull the correct image automatically with DOCKER_AUTO_PULL=0)
```

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:42]: For some reason this bails out on my system with

    /home/mikolai/TUD/Code/RIOT/makefiles/docker.inc.mk:42: *** target pattern contains no '%'.  Stop.

Seems to be connected to the usage of `@`.

Comment (maribu) [reply makefiles/docker.inc.mk]: That error message is pretty crazy, as if make was parsing the arguments to `$(shell ...)` as a make target.

Did quoting it help?

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:44]:
```suggestion
        $(info Building with latest available riotbuild image. You can pull the correct image automatically with DOCKER_AUTO_PULL=1)
```

Sorry, you were too fast :P

File: makefiles/docker.inc.mk
Comment (maribu) [reply makefiles/docker.inc.mk]: Which version of make are you using btw.? The one from latest Ubuntu LTS?

Comment (mguetschow) [reply makefiles/docker.inc.mk]: Humm I tried

```
$(shell $(DOCKER) pull "$(DOCKER_REPO)@sha256:$(DOCKER_TESTED_IMAGE_REPO_DIGEST)")
```

still the same :/

This is `GNU Make 4.3`.

Comment (maribu) [reply makefiles/docker.inc.mk]: And now?

Comment (maribu) [reply makefiles/docker.inc.mk]: In a debian podman container I ran:

```
$ DOCKER=podman make BUILD_IN_DOCKER=1
/home/maribu/Repos/software/RIOT/master/makefiles/docker.inc.mk:39: Required docker image sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b3 not installed
Pulling required image now automatically. You can disable this with DOCKER_AUTO_PULL=0
Trying to pull docker.io/riot/riotbuild@sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1...
Getting image source signatures
Copying blob 641b60a8ef3b done  
```

In that container, I also have

```
$ make --version
GNU Make 4.3
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

Comment (mguetschow) [reply makefiles/docker.inc.mk]: Fun, works for me too with `DOCKER=podman`, but not with `DOCKER=docker` (or no env variable). :confused:

Comment (mguetschow) [reply makefiles/docker.inc.mk]: I'm really out of ideas on how to debug this... :/

File: makefiles/docker.inc.mk
Comment (mguetschow) [inline makefiles/docker.inc.mk:35]: Just noticed that this is executed even with `BUILD_IN_DOCKER=0` (just adding a second condition to the if would help).

Comment (maribu) [reply makefiles/docker.inc.mk]: Good catch, fixed

File: makefiles/docker.inc.mk
Comment (maribu) [reply makefiles/docker.inc.mk]: I sadly cannot install docker in a container. I did try it again with a debian stable and podman-docker (that is, podman provides the docker command). That did just work. (And according to the error message the error should occur prior to podman or docker being invoked.)

Could you check if `env -i PATH="$PATH" make BUILD_IN_DOCKER=1` works for you? Maybe there is some environment variable that somehow affects the process that I do not have in my freshly set up debian container?

Comment (mguetschow) [reply makefiles/docker.inc.mk]: > Could you check if `env -i PATH="$PATH" make BUILD_IN_DOCKER=1` works for you? Maybe there is some environment variable that somehow affects the process that I do not have in my freshly set up debian container?

Sadly still the same problem.

Comment (mguetschow) [reply makefiles/docker.inc.mk]: FWIW, I can reproduce with a minimal example: `make all` on this Makefile:

```
$(shell docker pull docker.io/riot/riotbuild@sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1)

all:
```

**Comment by kfessel** [comment]
it seems like digest needs to be  `sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1`

to receive `sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b3`

```
$ ./dockery.sh riot/riotbuild manifests/latest
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 78827,
    "digest": "sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b3"
  },
  ...
```
```
$./dockery.sh riot/riotbuild manifests/latest
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 78827,
    "digest": "sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b3"
  },
 ```
```
$ ./dockery.sh riot/riotbuild manifests/sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 78827,
    "digest": "sha256:f5951bc41dfface6cac869181d703e62cbdd3b7976b0946130a38f2e658000b3"
  },
  
  ...
```

dockery.sh  is for dancing around that token thing that docker hub put in front of the [OCI container  API](https://github.com/opencontainers/distribution-spec/blob/v1.0.1/spec.md)

it is a stripped version of this 

https://stackoverflow.com/questions/57316115/get-manifest-of-a-public-docker-image-hosted-on-docker-hub-using-the-docker-regi
``` 
#!/bin/sh

repo=$1
q=$2
api="application/vnd.docker.distribution.manifest.v2+json"
apil="application/vnd.docker.distribution.manifest.list.v2+json"
token=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:${repo}:pull" \
        | jq -r '.token')
curl -H "Accept: ${api}" -H "Accept: ${apil}" \
     -H "Authorization: Bearer $token" \
     -s "https://registry-1.docker.io/v2/${repo}/${q}" | jq .

```

File: makefiles/docker.inc.mk
Comment (maribu) [reply makefiles/docker.inc.mk]: Does it work if you do

```Makefile
IGNORED := $(shell ...)
```

instead of

```Makefile
$(shell ...)
```

?

Comment (mguetschow) [reply makefiles/docker.inc.mk]: This however works:

```
TEST := $(shell docker pull docker.io/riot/riotbuild@sha256:75dec511ba26424987a26bdee5ac2f94d5f4928d79b627d1620b9d2391aab3e1)

all:
```

Comment (maribu) [reply makefiles/docker.inc.mk]: I just added the `docker-pull` make goal and add this as dependency to the command running docker if and only if the image is locally not available and automatic pulling is enabled.

**Comment by maribu** [comment]
@kfessel Yes. This is already implemented in this PR.

File: makefiles/docker.inc.mk
Comment (mguetschow) [reply makefiles/docker.inc.mk]: That works for me! :tada:

**Review by mguetschow — APPROVED** [review]
LGTM then, thanks for your effort! :hugs:

But let's hold this back until #20721  is merged.

**Comment by maribu** [comment]
Thanks a bunch! 😄 🎉

**Pull request merged**
